File: material/algorithm/sort/InsertSort.java
package material.algorithm.sort;

import java.util.Arrays;

public class InsertSort {
  public static void main(String[] args) {
    int[] arr = {412, 34, 31, 32, 12};
    sort(arr);
    System.out.print(Arrays.toString(arr));
  }
  public static void sort(int[] arr) {
    for (int i = 1; i < arr.length; i++) {
      int value = arr[i];
      int j = i - 1;
      for (; j >=0; j--) {
        if (arr[j] > value) {
          arr[j + 1] = arr[j];
        } else {
          break;
        }
      }
      // 这里是j+1;因为最后还要执行一次j--
      arr[j + 1] = value;
    }
  } 
}

File: material/algorithm/sort/InsertSort.js
let arr = [412, 34, 31, 32, 12]

function insertSort (arr) {
  if (arr.length <= 1) return
  for (let i = 1; i < arr.length; i++) {
    let value = arr[i]
    let j = i - 1
    // 为什么叫插入排序？
    // 下面的循环的意思只要第i位比前面的任意一位(j)小，将j至i-1位向后移动一位。
    // 然后将第i位的值放入到最后一个比i大的位置。
    // 注意，由于上面插入操作，i前的所有位都是按照从小到大排序了。
    //（有点人叫“动态排序”，即动态地往有序集合中添加数据）
    for (; j >= 0; j--) {
      if (arr[j] > value) {
        arr[j + 1] = arr[j];
      } else {
        break
      }
    }
    arr[j + 1] = value
  }

  return arr
}

insertSort(arr)